satellite/contact/endpoint.go

```diff
@@ -28,10 +34,87 @@ func NewEndpoint(log *zap.Logger, service *Service) *Endpoint {
 // Checkin is periodically called by storage nodes to keep the satellite informed of its existence,
 // address, and operator information. In return, this satellite keeps the node informed of its
 // reachability.
+// When a node checkins with the satellite, the satellite pings the node back to confirm they can
+// successfully connect.
 func (endpoint *Endpoint) Checkin(ctx context.Context, req *pb.CheckinRequest) (_ *pb.CheckinResponse, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	// TODO: record information, pingback node here
+	peerID, err := peerIDFromContext(ctx)
+	pingNodeSuccess, pingErrorMessage, err := pingBack(ctx, endpoint, req, peerID)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	// TODO(jg): We are making 2 requests to the database, one to update uptime and
+	// the other to update the capacity and operator info. We should combine these into
+	// one to reduce db connections. Consider adding batching and using a stored procedure.
+	_, err = endpoint.service.overlay.UpdateUptime(ctx, peerID, pingNodeSuccess)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	nodeInfo := pb.InfoResponse{Operator: req.GetOperator(), Capacity: req.GetCapacity()}
+	_, err = endpoint.service.overlay.UpdateNodeInfo(ctx, peerID, &nodeInfo)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	return &pb.CheckinResponse{
+		PingNodeSuccess:  pingNodeSuccess,
+		PingErrorMessage: pingErrorMessage,
+	}, nil
+}
+
+func pingBack(ctx context.Context, endpoint *Endpoint, req *pb.CheckinRequest, peerIDFromContext storj.NodeID) (bool, string, error) {
+	client, err := newClient(ctx,
+		endpoint.log,
+		endpoint.service.transport,
+		req.GetAddress(),
+	)
+	if err != nil {
+		// if this is a network error, then return the error otherwise just report internal error
+		_, ok := err.(net.Error)
+		if ok {
+			return false, "", Error.New("couldn't connect to client at addr %s due to network error: %v.", req.GetAddress().String(), err)
+		}
+		return false, "", Error.New("couldn't connect to client at addr: %s due to internal error.", req.GetAddress().String())
+	}
+
+	pingNodeSuccess := true
+	var pingErrorMessage string
 
-	return &pb.CheckinResponse{}, nil
+	p := &peer.Peer{}
+	_, err = client.pingNode(ctx, &pb.ContactPingRequest{}, grpc.Peer(p))
+	if err != nil {
```

Comment:
If there is an error, should we return it here rather than continuing down and checking the identity?